### PR TITLE
make builds deterministic by default

### PR DESCRIPTION
**Pull request by alessio**
---
name: Pull Request
about: Submit a pull request
title: 'make builds deterministic by default'
labels: ''
assignees: '@boz @gosuri'

---

## Description

Unless nostrip is passed via BUILD_OPTIONS, builds should
be deterministic by default.

## Motivation and Context

This is a dependency of the work I'm doing to try integrate
Akash's Makefile with the Cosmos SDK's new simplified
deterministic buildsystem.

Thanks for considering.

## How Has This Been Tested?

* `make akash BUILD_OPTIONS=nostrip`
* `make install BUILD_OPTIONS=nostrip`
* `make akash`
* `make install`

## Types of changes

- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)

## Checklist:
- [ ] My code follows the code style of this project.
- [x] My change requires a change to the documentation.
- [ ] I have updated the documentation accordingly.
- [ ] I've read the [CONTRIBUTION](https://github.com/ovrclk/akash/blob/master/CONTRIBUTING.md) guide
- [x] I have signed-off my commits with `git commit -s`
- [ ] I have added tests to cover my changes.
- [ ] All new and existing tests passed.

### CODE REVIEW

**Comment by troian** [comment]
Thanks for the PR, a comment inside

File: Makefile [hunk 1]
```diff
@@ -54,11 +54,24 @@ GORELEASER_LD_FLAGS = '-s -w -X github.com/cosmos/cosmos-sdk/version.Name=akash
...
+ldflags = -X github.com/cosmos/cosmos-sdk/version.Name=akash \
 -X github.com/cosmos/cosmos-sdk/version.AppName=akash \
 -X "github.com/cosmos/cosmos-sdk/version.BuildTags=$(BUILD_TAGS)" \
 -X github.com/cosmos/cosmos-sdk/version.Version=$(shell git describe --tags | sed 's/^v//') \
--X github.com/cosmos/cosmos-sdk/version.Commit=$(shell git log -1 --format='%H')'
+-X github.com/cosmos/cosmos-sdk/version.Commit=$(shell git log -1 --format='%H')
+
+# check for nostrip option
+ifeq (,$(findstring nostrip,$(BUILD_OPTIONS)))
+  ldflags += -s -w
```
[the remark this answers is not in the document]
Comment (alessio) [reply Makefile:65]: Yes, I could! I may need your help though, as I'm not sure about how I could test my changes.

[the remark this answers is not in the document]
Comment (troian) [reply Makefile:65]: `make release-dry-run` is essentially goreleaser wrapped into docker with some cross-compile magic. it produces artifacts into `dist` directory. would this help?

[the remark this answers is not in the document]
Comment (alessio) [reply Makefile:65]: Am I wrong in thinking that you may want your binaries to be released `stripped` anyway? As in: `goreleaser` could just have the `-trimpath` and the `-s -w` ldflags appended to the command line's args unconditionally.

[the remark this answers is not in the document]
Comment (troian) [reply Makefile:65]: probably, we still need `-s -w` to shrink debug information though

[the remark this answers is not in the document]
Comment (alessio) [reply Makefile:65]: You seem to already have that:

```make
GORELEASER_LD_FLAGS = '-s -w -X github.com/cosmos/cosmos-sdk/version.Name=akash \
-X github.com/cosmos/cosmos-sdk/version.AppName=akash \
-X "github.com/cosmos/cosmos-sdk/version.BuildTags=$(GORELEASER_BUILD_TAGS)" \
-X github.com/cosmos/cosmos-sdk/version.Version=$(shell git describe --tags --abbrev=0) \
-X github.com/cosmos/cosmos-sdk/version.Commit=$(shell git log -1 --format='%H')'
```

You only lack `-trimpath` AFAICS

[the remark this answers is not in the document]
Comment (troian) [reply Makefile:65]: ah, sorry, though you suggesting to remove `-s -w` from goreleaser
yah, this way should be good `GORELEASER_LD_FLAGS = '-trimpath -s -w`

[the remark this answers is not in the document]
Comment (alessio) [reply Makefile:65]: I've appended `-trimpath` to `GORELEASER_FLAGS` as it needs to be passed on the `go build` arguments pipeline directly (and not via `-ldflags`). Please see 790b52fa9a2d4ad90edcde4f2be39320481b65e7

[the remark this answers is not in the document]
Comment (troian) [reply Makefile:65]: sorry for bothering, i just released from [this error](https://github.com/ovrclk/akash/pull/846/checks?check_run_id=1229604953) it won't work like that. we need to change that in goreleaser files. could you remove usage of `-trimpath` from goreleaser flags. we will make separate PR

File: Makefile [hunk 2]
```diff
@@ -47,18 +47,31 @@ else
 	GORELEASER_BUILD_TAGS=$(BUILD_TAGS),testnet
 endif
 
-GORELEASER_FLAGS    = -tags="$(GORELEASER_BUILD_TAGS)"
+GORELEASER_FLAGS    = -trimpath -tags="$(GORELEASER_BUILD_TAGS)"
```
Comment (alessio) [inline Makefile:50]:
```suggestion
GORELEASER_FLAGS    = -tags="$(GORELEASER_BUILD_TAGS)"
```

File: Makefile (hunk 1 above)
[the remark this answers is not in the document]
Comment (alessio) [reply Makefile:65]: Reverted

**Review by troian — APPROVED** [review]

**Pull request merged**